File: interviews/lc/06-array-intersection.py
from typing import List
import time


def intersectHash(nums1: List[int], nums2: List[int]) -> List[int]:
    nums1Freqs = {}
    for n in nums1:
        if n in nums1Freqs:
            nums1Freqs[n] += 1
        else:
            nums1Freqs[n] = 1
    result = []
    for n in nums2:
        if n in nums1Freqs and nums1Freqs[n] > 0:
            result.append(n)
            nums1Freqs[n] -= 1
    return result


def intersectSort(nums1: List[int], nums2: List[int]) -> List[int]:
    nums1.sort()
    nums2.sort()
    ptr1Idx = 0
    ptr2Idx = 0
    result = []
    while ptr1Idx < len(nums1) and ptr2Idx < len(nums2):
        n1 = nums1[ptr1Idx]
        n2 = nums2[ptr2Idx]
        if n1 == n2:
            result.append(n1)
            ptr1Idx += 1
            ptr2Idx += 1
        elif n1 < n2:
            ptr1Idx += 1
        else:  # n1 > n2
            ptr2Idx += 1
    return result


def runTestCase(nums1: List[int], nums2: List[int]) -> None:
    nums1Str = ",".join([str(n) for n in nums1])
    nums2Str = ",".join([str(n) for n in nums2])
    start = time.perf_counter_ns()
    result = intersectHash(nums1, nums2)
    end = time.perf_counter_ns()
    resultStr = ",".join([str(n) for n in result])
    delta = end - start
    print(f"intersectHash: {nums1Str} & {nums2Str} = {resultStr} {delta}ns")

    nums1Str = ",".join([str(n) for n in nums1])
    nums2Str = ",".join([str(n) for n in nums2])
    start = time.perf_counter_ns()
    result = intersectSort(nums1, nums2)
    end = time.perf_counter_ns()
    resultStr = ",".join([str(n) for n in result])
    delta = end - start
    print(f"intersectSort: {nums1Str} & {nums2Str} = {resultStr} {delta}ns")


if __name__ == "__main__":
    runTestCase([1], [1])
    runTestCase([1], [2])
    runTestCase([1, 2, 2, 1], [2, 2])
    runTestCase([4, 9, 5, 4], [9, 4, 9, 8, 4])

File: interviews/lc/06-array-intersection.cpp
#include "lc-utility.h"

vector<int> intersectHash(const vector<int>& nums1, const vector<int>& nums2) {
  unordered_map<int, int> num1Freqs;
  for (const auto n : nums1) {
    num1Freqs[n]++;
  }
  vector<int> result;
  for (const auto n : nums2) {
    if (num1Freqs.count(n) > 0 && num1Freqs[n] > 0) {
      result.push_back(num1Freqs[n]);
      num1Freqs[n]--;
    }
  }
  return result;
}

vector<int> intersectSort(vector<int> nums1, vector<int> nums2) {
  std::sort(nums1.begin(), nums1.end());
  std::sort(nums2.begin(), nums2.end());
  int ptr1 = 0;
  int ptr2 = 0;
  vector<int> result;
  while (ptr1 < nums1.size() && ptr2 < nums2.size()) {
    const auto n1 = nums1[ptr1];
    const auto n2 = nums2[ptr2];
    if (n1 == n2) {
      result.push_back(n1);
      ptr1++;
      ptr2++;
    } else if (n1 < n2) {
      ptr1++;
    } else {  // n1 > n2
      ptr2++;
    }
  }
  return result;
}

void runTestCase(const vector<int>& nums1, const vector<int>& nums2) {
  const auto nums1Str = vector2str(nums1);
  const auto nums2Str = vector2str(nums2);
  {
    const auto start = high_resolution_clock::now();
    const auto result = intersectHash(nums1, nums2);
    const auto end = high_resolution_clock::now();
    const auto resultStr = vector2str(result);
    const auto runtimeNs = nanoseconds(end - start).count();
    cout << "intersectHash: " << nums1Str << " & " << nums2Str << " => "
         << resultStr << runtimeNs << "ns" << endl;
  }
  {
    const auto start = high_resolution_clock::now();
    const auto result = intersectSort(nums1, nums2);
    const auto end = high_resolution_clock::now();
    const auto resultStr = vector2str(result);
    const auto runtimeNs = nanoseconds(end - start).count();
    cout << "intersectSort: " << nums1Str << " & " << nums2Str << " => "
         << resultStr << runtimeNs << "ns" << endl;
  }
}

int main(int argc, char* argv[]) {
  runTestCase({1, 2, 2, 1}, {2, 2});
  runTestCase({4, 9, 5}, {9, 4, 9, 8, 4});
  return EXIT_SUCCESS;
}